feat: enable response type infer for API routes with params

didavid61202 opened the pull request:
<!---
☝️ Please ensure title is following conventional commits (https://conventionalcommits.org)

Examples:
 - feat: add something
 - fix(rollup): change something
 - docs: fix typo
-->

### 🔗 Linked issue

resolves #221
<!-- Please ensure there is an open issue and mention its number as #123 -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description
Update `MatchedRoutes` helper type in `types/fetch.ts` to enable  `$fetch` to infer response type even for routes with params, (ex: users/[useId]/posts/[postId].ts)

This will provide better DX for a wider range of complex requests with variables in it for requesting server routes containing params

This will also future provide Nuxt 3's `useFetch` and `useLazyFetch` to be able to infer the response's data type. 😎 

### Usage scenario

#### assume the project contains  `routes\users\[userId].ts` and `routes\users\[userId]\posts\[postId].ts`
```ts
// routes\users\[userId].ts
export default eventHandler(() => {
    return { username: 'David', age: 18 }
})
```
```ts
// routes\users\[userId]\posts\[postId].ts
export default eventHandler(() => {
    return { content: 'post content', author: 'David', points: 99 }
})
```
#### and use in other file as 
```ts
declare const params
cosnt { userId, postId } = params

const user = await $fetch(`/users/${userId}`) // or $fetch('/users/david') 
// user's type will be 
// { 
//    username: string; 
//    age: number; 
// }

const post = await $fetch(`/users/${userId}/posts/${postId}`) // or $fetch('/users/david/posts/42')
// post's type will be 
// { 
//    content: string; 
//    author: string; 
//    points: number; 
// }

const result = await $fetch(`/users/${userId}/notInServerRoutes/${postId}`)
// result's type will be `unknown`

```
<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it resolves an open issue, please link to the issue here. For example "Resolves #1337" -->

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

didavid61202 commented:
btw, I've asked in Nuxt's Dicord/typescript room [about playgound/tsconfig.json setup](https://discordapp.com/channels/473401852243869706/496688443141652500/973673438080286761) for correctly inferring nitro's `InternalApi` type during dev.

danielroe commented:
Would also be nice to support globs, such as `/api/foo/**`. Do you think you would be able to add that @didavid61202?

didavid61202 commented:
> Would also be nice to support globs, such as `/api/foo/**`. Do you think you would be able to add that @didavid61202?

on it! 👌

didavid61202 commented:
@danielroe Hi! 👋  
I've updated and refactored the `MatchRoutes` type to support globs `user/id/**`. And improved the matching result by using different priorities of exact, partial, exact + partial, template string/const string matches.

I also added some type tests using the `expect-type` package.

please check the updated [PR description](https://github.com/unjs/nitro/pull/222#issue-1231682810) or the [type tests code](https://github.com/unjs/nitro/blob/9579f40f8006afe40900539a854cf72347431212/test/fixture/types.ts) for details. 🙌 

Thanks for your time and effort working on this and guiding me. ☺️

danielroe commented:
This is looking great - just want to clarify what I meant by globs.

If we have a key in `InternalApi` like this:

```ts
export interface InternalApi {
  '/api/user/:userId/post/**': SomeReturnType
}
```

Then this should match `'/api/user/234/post' and also '/api/user/2348/post/parent' and '/api/user/2348/post/parent/child'.

didavid61202 commented:
> This is looking great - just want to clarify what I meant by globs.
> 
> If we have a key in `InternalApi` like this:
> 
> ```ts
> export interface InternalApi {
>   '/api/user/:userId/post/**': SomeReturnType
> }
> ```
> 
> Then this should match `/api/user/234/post` and also `/api/user/2348/post/parent` and `/api/user/2348/post/parent/child`.

Got it, working on it 👌

didavid61202 commented:
Implemented globs 🎉 
please check if it's working as expected at your earliest convenience, Thanks ☺️ 

I think [this test case](https://github.com/unjs/nitro/blob/966708eca47fe93d06e382ff901557bc6f29a7e0/test/fixture/types.ts#L84) should also include Api key: `/api/todos/:todoId/comments/**:commentId`, @danielroe what is your opinion ? 🤔

didavid61202 commented:
type check errors in `ci` check action while running `test:types` script can be fix by  
- #269

didavid61202 commented:
I've merge from main and `test:types` is passing now 🎉

Hi @danielroe, please review when you have time 😊

danielroe reviewed (approved):
This looks good to me. ✨

**Side note**: I'm a little apprehensive about potential type complexity but I'd be happy to merge and refactor type implementation down the line if needed. (For example, perhaps we could explicitly generate `/api/${string}` rather than `/api/:slug` as the key of `InternalApi` if we find the current approach has negative perf implications.)

pi0 commented:
Thanks @didavid61202 @danielroe <3
